src/main/java/com/powsybl/openloadflow/equations/EquationSystem.java

```diff
@@ -34,77 +34,89 @@
-        }
+            // index derivatives per variable then per equation
 
-        private void reIndex() {
-            sortedEquationsToSolve.clear();
-            sortedVariablesToFind.clear();
+            // equations to remove
+            for (Equation equation : equationsToRemove) {
+                sortedEquationsToSolve.remove(equation);
+                for (EquationTerm equationTerm : equation.getTerms()) {
+                    for (Variable variable : equationTerm.getVariables()) {
+                        Set<Equation> equationsUsingThisVariable = sortedVariablesToFind.get(variable);
+                        if (equationsUsingThisVariable != null) {
+                            equationsUsingThisVariable.remove(equation);
+                            if (equationsUsingThisVariable.isEmpty()) {
+                                sortedVariablesToFind.remove(variable);
+                            }
+                        }
+                    }
+                }
+            }
 
-            Set<Variable> variablesToFind = new HashSet<>();
-            for (Equation equation : equations.values()) {
+            // equations to add
+            for (Equation equation : equationsToAdd) {
                 if (equation.isActive() && EquationUpdateType.DEFAULT == equation.getUpdateType()) {
-                    // do not use equations that would be updated only after NR
-                    NavigableMap<Variable, List<EquationTerm>> equationTermsByVariable = null;
-                    // check we have at least one equation term active
-                    boolean atLeastOneTermIsValid = false;
                     for (EquationTerm equationTerm : equation.getTerms()) {
                         if (equationTerm.isActive()) {
-                            atLeastOneTermIsValid = true;
-                            if (equationTermsByVariable == null) {
-                                equationTermsByVariable = sortedEquationsToSolve.computeIfAbsent(equation, k -> new TreeMap<>());
-                            }
                             for (Variable variable : equationTerm.getVariables()) {
-                                equationTermsByVariable.computeIfAbsent(variable, k -> new ArrayList<>())
+                                sortedEquationsToSolve.computeIfAbsent(equation, k -> new TreeMap<>())
+                                        .computeIfAbsent(variable, k -> new ArrayList<>())
                                         .add(equationTerm);
-                                variablesToFind.add(variable);
+                                sortedVariablesToFind.computeIfAbsent(variable, k -> new TreeSet<>())
+                                        .add(equation);
                             }
                         }
                     }
-                    if (!atLeastOneTermIsValid) {
-                        throw new IllegalStateException("Equation " + equation + " is active but all of its terms are inactive");
```

Comment:
Oh I forgot about that, I didn't get why you removed that check, it could not happen anymore?